aggregator/generic_elem.go

```diff
@@ -277,9 +279,23 @@ func (e *GenericElem) Consume(
 	// Process the aggregations that are ready for consumption.
 	for i := range e.toConsume {
 		timeNanos := timestampNanosFn(e.toConsume[i].startAtNanos, resolution)
-		e.processValue(timeNanos, e.toConsume[i].lockedAgg, flushLocalFn, flushForwardedFn)
+		e.toConsume[i].lockedAgg.Lock()
+		e.processValueWithAggregationLock(timeNanos, e.toConsume[i].lockedAgg, flushLocalFn, flushForwardedFn)
 		// Closes the aggregation object after it's processed.
+		e.toConsume[i].lockedAgg.closed = true
 		e.toConsume[i].lockedAgg.aggregation.Close()
+		if e.toConsume[i].lockedAgg.sourcesSeen != nil {
+			sourceSetSize := len(e.toConsume[i].lockedAgg.sourcesSeen)
+			if sourceSetSize <= maxCachedSourceSetSize {
+				e.cachedSourceSetsLock.Lock()
+				if len(e.cachedSourceSets) < maxNumCachedSourceSets {
```

Comment:
what happens if len(e.cachedSourceSets) >= maxNumCachedSourceSets? why not keep appending?

Reply:
This is to make sure we don't take up too much space caching the source sets. Usually there will be two live source sets: one that's just finished and ready to be consumed, and the other that's being actively added to. Can tune more if needed.

Reply:
kk, maybe just add some comments for this? Also why is this logic only executed when `sourceSetSize <= maxCachedSourceSetSize`?

Reply:
Sure yeah can add comments.

In a similar vein, if the source set we are about to cache is too big (i.e., bigger than the configured max size), we do not want to retain it and instead simply discard it (this is useful if say over the course of a month the raw metric IDs producing the forward metric ID has changed significantly so there are a lot of stale ids in the map so we regenerate the map once in a while to refresh the IDs in the map).